core/trino-spi/src/main/java/io/trino/spi/block/BlockUtil.java

```diff
@@ -185,9 +185,9 @@ static int countUsedPositions(boolean[] positions)
     {
         int used = 0;
         for (boolean position : positions) {
-            if (position) {
-                used++;
-            }
+            // Avoid branching by casting boolean to integer.
```

Comment:
What evidence do we have that this is from branch mispredictions? Do we have a capture from JMH with the perf profiler that shows this?

Reply:
It's similar concept to: https://github.com/trinodb/trino/pull/8624
which was giving around 50% improvements in JMH:
```
Benchmark             Mode  Cnt      Score      Error  Units
BenchmarkFoo.newPos  thrpt   10  96730,342 ± 2154,841  ops/s
BenchmarkFoo.oldPos  thrpt   10  50875,222 ± 1058,018  ops/s
```
I don't think @raunaqmorarka did microbenchmarks on this one though

Reply:
Understood, but it’s unclear whether it’s because of branch mispredictions or some other effect.

Reply:
> Understood, but it’s unclear whether it’s because of branch mispredictions or some other effect.

I didn't look into assembly.